# Import Libraries as we go

In [72]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb
from sklearn.preprocessing import LabelEncoder
from sklearn.feature_selection import mutual_info_classif

In [56]:
df = pd.read_csv("data/fraudTest.csv")
df.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2020-06-21 12:14:25  2291163933867244   
1           1   2020-06-21 12:14:33  3573030041201292   
2           2   2020-06-21 12:14:53  3598215285024754   
3           3   2020-06-21 12:15:15  3591919803438423   
4           4   2020-06-21 12:15:17  3526826139003047   

                               merchant        category    amt   first  \
0                 fraud_Kirlin and Sons   personal_care   2.86    Jeff   
1                  fraud_Sporer-Keebler   personal_care  29.84  Joanne   
2  fraud_Swaniawski, Nitzsche and Welch  health_fitness  41.28  Ashley   
3                     fraud_Haley Group        misc_pos  60.05   Brian   
4                 fraud_Johnston-Casper          travel   3.19  Nathan   

       last gender                       street  ...      lat      long  \
0   Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1  Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2     Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3  Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4    Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   

   city_pop                     job         dob  \
0    333497     Mechanical engineer  1968-03-19   
1       302  Sales professional, IT  1990-01-17   
2     34496       Librarian, public  1970-10-21   
3     54767            Set designer  1987-07-25   
4      1126      Furniture designer  1955-07-06   

                          trans_num   unix_time  merch_lat  merch_long  \
0  2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1  324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2  c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3  2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4  57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

In [57]:
df.drop(columns="Unnamed: 0", inplace=True)
df.head()

trans_date_trans_time            cc_num  \
0   2020-06-21 12:14:25  2291163933867244   
1   2020-06-21 12:14:33  3573030041201292   
2   2020-06-21 12:14:53  3598215285024754   
3   2020-06-21 12:15:15  3591919803438423   
4   2020-06-21 12:15:17  3526826139003047   

                               merchant        category    amt   first  \
0                 fraud_Kirlin and Sons   personal_care   2.86    Jeff   
1                  fraud_Sporer-Keebler   personal_care  29.84  Joanne   
2  fraud_Swaniawski, Nitzsche and Welch  health_fitness  41.28  Ashley   
3                     fraud_Haley Group        misc_pos  60.05   Brian   
4                 fraud_Johnston-Casper          travel   3.19  Nathan   

       last gender                       street        city  ...      lat  \
0   Elliott      M            351 Darlene Green    Columbia  ...  33.9659   
1  Williams      F             3638 Marsh Union     Altonah  ...  40.3207   
2     Lopez      F         9333 Valentine Point    Bellmore  ...  40.6729   
3  Williams      M  32941 Krystal Mill Apt. 552  Titusville  ...  28.5697   
4    Massey      M     5783 Evan Roads Apt. 465    Falmouth  ...  44.2529   

       long  city_pop                     job         dob  \
0  -80.9355    333497     Mechanical engineer  1968-03-19   
1 -110.4360       302  Sales professional, IT  1990-01-17   
2  -73.5365     34496       Librarian, public  1970-10-21   
3  -80.8191     54767            Set designer  1987-07-25   
4  -85.0170      1126      Furniture designer  1955-07-06   

                          trans_num   unix_time  merch_lat  merch_long  \
0  2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1  324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2  c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3  2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4  57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 22 columns]

In [58]:
df.describe()

cc_num            amt            zip            lat  \
count  5.557190e+05  555719.000000  555719.000000  555719.000000   
mean   4.178387e+17      69.392810   48842.628015      38.543253   
std    1.309837e+18     156.745941   26855.283328       5.061336   
min    6.041621e+10       1.000000    1257.000000      20.027100   
25%    1.800429e+14       9.630000   26292.000000      34.668900   
50%    3.521417e+15      47.290000   48174.000000      39.371600   
75%    4.635331e+15      83.010000   72011.000000      41.894800   
max    4.992346e+18   22768.110000   99921.000000      65.689900   

                long      city_pop     unix_time      merch_lat  \
count  555719.000000  5.557190e+05  5.557190e+05  555719.000000   
mean      -90.231325  8.822189e+04  1.380679e+09      38.542798   
std        13.721780  3.003909e+05  5.201104e+06       5.095829   
min      -165.672300  2.300000e+01  1.371817e+09      19.027422   
25%       -96.798000  7.410000e+02  1.376029e+09      34.755302   
50%       -87.476900  2.408000e+03  1.380762e+09      39.376593   
75%       -80.175200  1.968500e+04  1.385867e+09      41.954163   
max       -67.950300  2.906700e+06  1.388534e+09      66.679297   

          merch_long       is_fraud  
count  555719.000000  555719.000000  
mean      -90.231380       0.003860  
std        13.733071       0.062008  
min      -166.671575       0.000000  
25%       -96.905129       0.000000  
50%       -87.445204       0.000000  
75%       -80.264637       0.000000  
max       -66.952026       1.000000

In [59]:
df[df.is_fraud==0].describe()

cc_num            amt            zip            lat  \
count  5.535740e+05  553574.000000  553574.000000  553574.000000   
mean   4.179643e+17      67.614408   48846.424285      38.541406   
std    1.309964e+18     152.471931   26855.813540       5.061194   
min    6.041621e+10       1.000000    1257.000000      20.027100   
25%    1.800429e+14       9.600000   26292.000000      34.668900   
50%    3.521417e+15      47.150000   48174.000000      39.371600   
75%    4.642255e+15      82.620000   72042.000000      41.894800   
max    4.992346e+18   22768.110000   99783.000000      65.689900   

                long      city_pop     unix_time      merch_lat  \
count  553574.000000  5.535740e+05  5.535740e+05  553574.000000   
mean      -90.230495  8.831369e+04  1.380683e+09      38.540954   
std        13.720111  3.006593e+05  5.203038e+06       5.095755   
min      -165.672300  2.300000e+01  1.371817e+09      19.027422   
25%       -96.798000  7.410000e+02  1.376031e+09      34.754055   
50%       -87.476900  2.408000e+03  1.380770e+09      39.374616   
75%       -80.175200  1.968500e+04  1.385872e+09      41.953397   
max       -67.950300  2.906700e+06  1.388534e+09      66.679297   

          merch_long  is_fraud  
count  553574.000000  553574.0  
mean      -90.230474       0.0  
std        13.731399       0.0  
min      -166.671575       0.0  
25%       -96.906402       0.0  
50%       -87.443645       0.0  
75%       -80.265911       0.0  
max       -66.952026       0.0

In [60]:
df[df.is_fraud==1].describe()

cc_num          amt           zip          lat         long  \
count  2.145000e+03  2145.000000   2145.000000  2145.000000  2145.000000   
mean   3.854274e+17   528.356494  47862.900233    39.019971   -90.445556   
std    1.276313e+18   392.747594  26706.283776     5.076671    14.147521   
min    6.048700e+10     1.780000   1257.000000    20.082700  -155.488000   
25%    3.056061e+13   214.510000  23937.000000    34.957200   -96.743000   
50%    2.358122e+15   371.940000  47987.000000    39.893600   -88.093500   
75%    4.335532e+15   907.770000  68031.000000    42.180800   -79.785600   
max    4.989848e+18  1320.920000  99921.000000    55.473200   -67.950300   

           city_pop     unix_time    merch_lat   merch_long  is_fraud  
count  2.145000e+03  2.145000e+03  2145.000000  2145.000000    2145.0  
mean   6.452932e+04  1.379587e+09    39.018618   -90.465222       1.0  
std    2.192075e+05  4.546697e+06     5.093782    14.159252       0.0  
min    2.300000e+01  1.371852e+09    19.161782  -156.480766       1.0  
25%    1.131000e+03  1.375756e+09    35.017480   -96.697100       1.0  
50%    2.870000e+03  1.379548e+09    39.744929   -88.040722       1.0  
75%    1.233500e+04  1.383350e+09    42.208725   -79.909134       1.0  
max    1.577385e+06  1.387754e+09    56.214113   -66.960745       1.0

As we can see, out of the 555719 records only 2145 are actually fraud, so we are dealing with the detection of a very minute data. We need to ensure high accuracy while also maintaining high precision.

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 555719 entries, 0 to 555718
Data columns (total 22 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   trans_date_trans_time  555719 non-null  object 
 1   cc_num                 555719 non-null  int64  
 2   merchant               555719 non-null  object 
 3   category               555719 non-null  object 
 4   amt                    555719 non-null  float64
 5   first                  555719 non-null  object 
 6   last                   555719 non-null  object 
 7   gender                 555719 non-null  object 
 8   street                 555719 non-null  object 
 9   city                   555719 non-null  object 
 10  state                  555719 non-null  object 
 11  zip                    555719 non-null  int64  
 12  lat                    555719 non-null  float64
 13  long                   555719 non-null  float64
 14  city_pop               555719 non-nu

# Removing the useless features and one hot encoding the discrete data

In [62]:
df.drop(columns=["trans_date_trans_time","first", "last", "trans_num", "unix_time", "dob", "street", "city", "state", "lat", "long"], inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 555719 entries, 0 to 555718
Data columns (total 11 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   cc_num      555719 non-null  int64  
 1   merchant    555719 non-null  object 
 2   category    555719 non-null  object 
 3   amt         555719 non-null  float64
 4   gender      555719 non-null  object 
 5   zip         555719 non-null  int64  
 6   city_pop    555719 non-null  int64  
 7   job         555719 non-null  object 
 8   merch_lat   555719 non-null  float64
 9   merch_long  555719 non-null  float64
 10  is_fraud    555719 non-null  int64  
dtypes: float64(3), int64(4), object(4)
memory usage: 46.6+ MB


In [66]:
le=LabelEncoder()

cat_column = df.select_dtypes(include=["object"]).columns.tolist()

for col in cat_column:
    df[col]=le.fit_transform(df[col])

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 555719 entries, 0 to 555718
Data columns (total 11 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   cc_num      555719 non-null  int64  
 1   merchant    555719 non-null  int32  
 2   category    555719 non-null  int32  
 3   amt         555719 non-null  float64
 4   gender      555719 non-null  int32  
 5   zip         555719 non-null  int64  
 6   city_pop    555719 non-null  int64  
 7   job         555719 non-null  int32  
 8   merch_lat   555719 non-null  float64
 9   merch_long  555719 non-null  float64
 10  is_fraud    555719 non-null  int64  
dtypes: float64(3), int32(4), int64(4)
memory usage: 38.2 MB


In [67]:
df.describe()

cc_num       merchant       category            amt  \
count  5.557190e+05  555719.000000  555719.000000  555719.000000   
mean   4.178387e+17     343.086430       6.237059      69.392810   
std    1.309837e+18     201.284993       3.912874     156.745941   
min    6.041621e+10       0.000000       0.000000       1.000000   
25%    1.800429e+14     166.000000       3.000000       9.630000   
50%    3.521417e+15     346.000000       6.000000      47.290000   
75%    4.635331e+15     515.000000      10.000000      83.010000   
max    4.992346e+18     692.000000      13.000000   22768.110000   

              gender            zip      city_pop            job  \
count  555719.000000  555719.000000  5.557190e+05  555719.000000   
mean        0.451367   48842.628015  8.822189e+04     241.755040   
std         0.497630   26855.283328  3.003909e+05     136.327924   
min         0.000000    1257.000000  2.300000e+01       0.000000   
25%         0.000000   26292.000000  7.410000e+02     124.000000   
50%         0.000000   48174.000000  2.408000e+03     239.000000   
75%         1.000000   72011.000000  1.968500e+04     362.000000   
max         1.000000   99921.000000  2.906700e+06     477.000000   

           merch_lat     merch_long       is_fraud  
count  555719.000000  555719.000000  555719.000000  
mean       38.542798     -90.231380       0.003860  
std         5.095829      13.733071       0.062008  
min        19.027422    -166.671575       0.000000  
25%        34.755302     -96.905129       0.000000  
50%        39.376593     -87.445204       0.000000  
75%        41.954163     -80.264637       0.000000  
max        66.679297     -66.952026       1.000000

In [68]:
df.corr()["is_fraud"]

cc_num       -0.001540
merchant     -0.000086
category      0.016980
amt           0.182267
gender        0.000748
zip          -0.002271
city_pop     -0.004910
job          -0.005074
merch_lat     0.005812
merch_long   -0.001060
is_fraud      1.000000
Name: is_fraud, dtype: float64

In [69]:
df.groupby("is_fraud")["amt"].describe()

count        mean         std   min     25%     50%     75%  \
is_fraud                                                                   
0         553574.0   67.614408  152.471931  1.00    9.60   47.15   82.62   
1           2145.0  528.356494  392.747594  1.78  214.51  371.94  907.77   

               max  
is_fraud            
0         22768.11  
1          1320.92

In [71]:
fraud_rate = df.groupby("category")["is_fraud"].mean().sort_values(ascending=False)
fraud_rate

category
11    0.012111
8     0.009756
4     0.009229
12    0.004278
2     0.002732
13    0.002292
3     0.002111
9     0.002082
10    0.001780
0     0.001471
5     0.001418
1     0.001375
7     0.001335
6     0.001280
Name: is_fraud, dtype: float64

In [73]:
mi = mutual_info_classif(df.drop("is_fraud", axis=1), df["is_fraud"])

In [74]:
mi

array([0.00691996, 0.00185236, 0.01114276, 0.01166397, 0.07283989,
       0.00756594, 0.00700367, 0.00429662, 0.00030383, 0.00026529])